### PR TITLE
Update Authorizer types and add initial docs

### PR DESCRIPTION
**Pull request by myleslinder**
I initially attempted to update the Remix dev dependencies but that caused two kinds of errors in the tests:
   1. `createCookieSessionStorage` is no longer exported directly from `@remix-run/server-runtime` but from `@remix-run/node` et al which was an easy fix. 
   2.  `new Request` with relative paths seemed to be failing with an invalid path error, however that isn't the typical behavior so I'm unsure of what the issue was. 
   
**Authorizer Updates**
- The only runtime code that was changed is the way [the default value to `raise` is assigned](https://github.com/myleslinder/remix-auth/blob/f63543987bf11da1f28a442bf3cfca85ec20defc/src/authorizer.ts#L70). This was required to get the types working properly and remove the expect errors.
  - If you'd prefer I not re-assign to the argument I can modify to use a local variable instead. 
  - The types should now be set up to infer the data property type from the first rule provided including if that rule is a route specific rule and there are no global rules.  
- The [context property of the `DataFunctionArgs` is made optional](https://github.com/myleslinder/remix-auth/blob/f63543987bf11da1f28a442bf3cfca85ec20defc/src/authorizer.ts#L12) to avoid having to provide the entire loader/action args if you don't use context in your app. Let me know if that doesn't work for you. 
- I've ran the tests and checked the types in the tests and everything seems to be working

**Docs Updates**
- I've put together an initial attempt at some bare-bones docs for `Authorizer` in the docs folder.

**Questions** 
- Is there a reason the rules are called sequentially as opposed to in parallel?
- There isn't really a good way to know which rule caused the authorization to fail. Would you be open to augmenting the existing `AuthorizationError` to include the rule name if applicable and used in the authorizer (instead of the generic `Error`)? Catching and checking for that error could make an easy way to set your own error messages for specific rules.

### CODE REVIEW

File: src/authorizer.ts [hunk 1]
```diff
@@ -1,20 +1,23 @@
...
-type AuthorizeArgs<User, Data> = Omit<RuleContext<User, Data>, "user">;
+/**
+ * Extra data passed to the Authorizer from the loader or action
+ */
+type AuthorizeArgs<Data> = { data?: Data } & Omit<
+  DataFunctionArgs,
+  "context"
+> & {
+    context?: DataFunctionArgs["context"];
+  };
```
Comment (sergiodxa) [inline src/authorizer.ts:13]: Context shouldn't be option, I know you may not use it, and it's the most common thing, but the type is not optional in LoaderArgs so it shouldn't be here, or at least not the types the rules receives, we could accept context as optional but pass it to the rules with a default empty object.

Comment (myleslinder) [reply src/authorizer.ts:13]: Makes sense. I've updated it to be optional in the call to `authorize` but the rule functions will always receive a context (an empty object if not provided).

File: src/authorizer.ts [hunk 2]
```diff
@@ -24,68 +27,61 @@ export interface RuleContext<User, Data = unknown> extends LoaderArgs {
...
-      failureRedirect,
-      raise = "response",
-      rules = [],
-    }: {
-      failureRedirect?: string;
-      raise?: "error" | "response" | "redirect";
-      rules?: RuleFunction<User, Data>[];
-    } = {}
   ): Promise<User> {
+    if (!raise) raise = "response";
```
Comment (sergiodxa) [inline src/authorizer.ts:70]: I'm thinking the default raise should be error, I think that's more expected and if the developer setup Sentry it should know if this throw and the dev is not catching it.

Also, Authorizer may receive the default raise and use it as default here

**Comment by sergiodxa** [comment]
> Is there a reason the rules are called sequentially as opposed to in parallel?

For consistency, the idea is that if the user doesn't pass the validation of two different rules the rule throwing should always be the same, the first one, and then after that's solved the second one, otherwise you will receive different errors every time you reload.

> There isn't really a good way to know which rule caused the authorization to fail. Would you be open to augmenting the existing AuthorizationError to include the rule name if applicable and used in the authorizer (instead of the generic Error)? Catching and checking for that error could make an easy way to set your own error messages for specific rules.

This sounds like a good idea.

**Comment by myleslinder** [comment]
Since it's really just about being able to provide specific error messages when certain rules fail, how would you feel about instead allowing the rules provided to be either a `RuleFunction` or an object with a `RuleFunction` and a message property, like so:

```ts
interface RuleFunction<User, Data> {
  (context: RuleContext<User, Data>): Promise<boolean>;
}
type Rule<User, Data> =
  | {
      message: string;
      rule: RuleFunction<User, Data>;
    }
  | RuleFunction<User, Data>;
```

It wouldn't break any existing use cases and would not be a breaking change. It would mean a small change to the checking of rules to something like:

```ts
for (let rule of [...this.rules, ...rules]) {
    let callableRule = typeof rule === "function" ? rule : rule.rule;
    let message =
      typeof rule === "function"
        ? `Forbidden${rule.name ? ` by policy ${rule.name}` : ""}`
        : rule.message;

    if (await callableRule({ user, ...args, context: args.context ?? {} })) continue;
    // omitted
```
> I have the real version of this implemented locally with some additional tests and all the previous tests continue to pass

Then if you need a custom error message for a rule you can just use the object version of the rule like so:

```ts
await authorizer.authorize(
  { request, params: { id: "1" } },
  {
    rules: [
      {
        rule: async ({ user }) => user.role !== "admin",
        message: "some message",
      },
    ],
  }
);
```
This is probably simpler than my initially proposed alternative which would result in being able to only use the error raise option, not being able to use arrow functions, having to wrap the call to authorize in a try/catch, check the thrown error is instanceof AuthorizationError and then check if a string value is equal to the name of one of your rule functions.

**Comment by sergiodxa** [comment]
I ended up dropping the Authorizer class, I think this needs DB access or a custom per app implementation to work well.

**Pull request closed without merging**